akka-javasdk/src/main/scala/akka/javasdk/impl/workflow/WorkflowImpl.scala

```diff
@@ -28,389 +33,267 @@ import akka.javasdk.impl.workflow.WorkflowEffectImpl.NoReply
+      transition match {
+        case StepTransition(stepName, input) =>
+          new SpiWorkflow.StepTransition(stepName, input.map(serializer.toBytes))
+        case Pause        => SpiWorkflow.Pause
+        case NoTransition => SpiWorkflow.NoTransition
+        case End          => SpiWorkflow.End
+      }
+
+    def handleState(persistence: Persistence[Any], transition: Transition): SpiWorkflow.Persistence =
+      persistence match {
+        case UpdateState(newState) =>
+          router._internalSetInitState(newState, transition.isInstanceOf[End.type])
+          new SpiWorkflow.UpdateState(serializer.toBytes(newState))
+        case DeleteState   => SpiWorkflow.DeleteState
+        case NoPersistence => SpiWorkflow.NoPersistence
+      }
+
+    effect match {
+      case error: ErrorEffectImpl[_] =>
+        new SpiWorkflow.Effect(
+          userState = SpiWorkflow.NoPersistence, // mean runtime don't need to persist any new state
+          SpiWorkflow.NoTransition,
+          reply = None,
+          error = Some(new SpiEntity.Error(error.description)))
+
+      case WorkflowEffectImpl(persistence, transition, reply) =>
+        val replyOpt = reply match {
+          case ReplyValue(value, _) => Some(value) // FIXME: Metadata is being discarded
+          case NoReply              => None
+        }
+
+        new SpiWorkflow.Effect(
+          handleState(persistence, transition), // can be null when fallback to emptyState
+          toSpiTransition(transition),
+          reply = replyOpt.map(serializer.toBytes),
+          error = None)
+
+      case TransitionalEffectImpl(persistence, transition) =>
+        new SpiWorkflow.Effect(
+          handleState(persistence, transition), // can be null when fallback to emptyState
+          toSpiTransition(transition),
+          reply = None,
+          error = None)
+    }
   }
 
-  private def runWorkflow(
-      init: WorkflowEntityInit): (Flow[WorkflowStreamIn, WorkflowStreamOut, NotUsed], WorkflowStreamOut) = {
-    val service =
-      services.getOrElse(init.serviceName, throw ProtocolException(init, s"Service not found: ${init.serviceName}"))
-    val router: WorkflowRouter[_, _] =
-      service.createRouter(new WorkflowContextImpl(init.entityId))
-    val workflowId = init.entityId
-
-    val workflowConfig =
-      WorkflowStreamOut(
-        WorkflowStreamOut.Message.Config(toWorkflowConfig(router._getWorkflowDefinition(), service.serializer)))
-
-    init.userState match {
-      case Some(state) =>
-        val bytesPayload = AnySupport.toSpiBytesPayload(state)
-        val decoded = service.serializer.fromBytes(bytesPayload)
-        router._internalSetInitState(decoded, init.finished)
-      case None => // no initial state
+  override def handleCommand(
+      userState: Option[BytesPayload],
+      command: SpiEntity.Command): Future[SpiWorkflow.Effect] = {
+
+    // can be null when fallback to emptyState
+    val decodedState = decodeState(userState)
```

Comment:
Does this handle the null case for emptyState? I think null was converted to BytesPayload.empty.

Reply:
Oh, no really true. First transition was setting the state already.

Reply:
Ok, it works for some other reason. The runtime is not using the emptyState at all. 

When it receives the first command, if passes `None` to the SPI.